---
title: "Pairwise differential expression"
author: "Kimberly Olney, Ph.D."
date: "2024-03-22"
output:
  html_document:
    df_print: paged
  pdf_document: default
params:
  args: myarg
---

# Project overview
Bulk mouse brain tissue samples treated with control, aducanumab, or Fc-dead aducanumab (3/group)

```{r setup, include=FALSE}
knitr::opts_chunk$set(root.dir = ".", echo = TRUE, warning=FALSE, message=FALSE)
```

# User defined variables
```{r set_variables}
source(here::here("bulkRNA/scripts/file_paths_and_colors.R"))
```

# Add sequence information to metadata 
```{r seq_info}
seq_info <- readLines("sample_read_info.txt")

# vectors to store extracted information
sample_ID <- vector()
instrument <- vector()
run <- vector()
flowcell <- vector()
lane <- vector()

# Loop through to obtain sequence information
for (line in seq_info) {
  # Extract Sample ID
  sample_ID <- c(sample_ID, strsplit(line, "/")[[1]][2])
  
  # Extract Instrument, Run, Flowcell, and Lane
  parts <- unlist(strsplit(line, " "))
  instrument_info <- unlist(strsplit(parts[1], "@"))
  instrument_info <- unlist(strsplit(instrument_info[2], ":"))
  instrument <- c(instrument, instrument_info[1])
  run_info <- unlist(strsplit(parts[1], ":"))
  run <- c(run, run_info[2])
  flowcell <- c(flowcell, run_info[3])
  lane <- c(lane, run_info[4])
}

# Create seq dataframe
seq_df <- data.frame(
  Short.ID = sample_ID,
  Instrument = instrument,
  Run = run,
  Flowcell = flowcell,
  Lane = lane
)

# Merge with metadata by Short.ID
metadata_updated <- merge(metadata, seq_df, by = "Short.ID")
write.table(metadata_updated, "../metadata_with_seq_info.tsv", quote = FALSE)
metadata <- metadata_updated

# Clean up
rm(metadata_updated, seq_df, flowcell, i, instrument, instrument_info, lane, line, parts, run, run_info, sample_ID, seq_info)
```

# Read in gene annotation file
```{r gene_annotation}
gtf.file <- paste0(pathToRef, "genes.gtf")
genes.gtf <- rtracklayer::import(gtf.file)
genes.gtf <- as.data.frame(genes.gtf)
genes.gtf <- genes.gtf[genes.gtf$type == "gene",]
rm(gtf.file)
```

# Read in counts data
```{r read_data}
# example of file naming:
# I_36_STAR.bamReadsPerGene.out.tab
# path to counts files
count_files <-
  file.path(paste0(
    "../starAligned/",
    metadata$Short.ID,
    typeOfCount
  ))
# add sample name to counts files
names(count_files) <- paste0(metadata$Short.ID)
# add sample count path information to metadata
metadata$path <- count_files
# make sure there is only unique paths
```

# Create a counts matrix
```{r counts_matrix}
counts <- data.frame(fread(count_files[1]))[c(1, 4)]
# Loop and read the 4th column remaining files
for (i in 2:length(count_files)) {
  counts <- cbind(counts, data.frame(fread(count_files[i]))[4])
}
rm(count_files)
```

# Star mapping metrics
```{r mapping_metrics}
star_metrics <- data.frame(counts[counts$V1 %like% "N_", ])
star_metrics$V1 <- as.vector(star_metrics$V1)
melt_star_metrics <- reshape::melt(star_metrics, id = c("V1"))
ggplot(melt_star_metrics, aes(x = value, fill = V1)) + 
  geom_histogram(position = "identity", bins = 100) + 
  facet_grid(V1 ~ .)  + theme_bw()
saveToPDF("../results/all_samples/library/mapping_metrics.pdf", width = 6, height = 7)
rm(star_metrics, melt_star_metrics)
```


# Format counts table
```{r counts_table}
# remove star metric information
counts <- counts[!grepl("N_", counts$V1),]
# set gene_id (i.e. V1) as the row name
row.names(counts) <- counts$V1
counts$V1 <- NULL
# set column names to NPID sample ID
colnames(counts) <- metadata$Short.ID
```

# Check inputs
```{r check_inputs}
all.equal(rownames(counts), genes.gtf$gene_id)
all.equal(colnames(counts), (metadata$Short.ID))

# rownames - gene_id to gene_name
all.equal(rownames(counts), genes.gtf$gene_id)
# how many duplicate gene_names are there?
table(duplicated(genes.gtf$gene_name))
# make unique gene names for duplicated gene names 
rownames(counts) <- make.unique(genes.gtf$gene_name)
genes.gtf$gene_name_unique <- rownames(counts)
```

# Create DGE object
```{r DGE_object}
# create object
dge <- DGEList(counts = counts,
               samples = metadata,
               genes = genes.gtf)

table(dge$samples$Treatment)
dge$samples$Treatment <- factor(dge$samples$Treatment, levels = c("Adu",  "Fc", "IgG"))

rm(counts)
```

# Save dge R object
```{r rOjects}
saveRDS(dge, file = "../rObjects/dge.raw.all.samples.rds")
#dge <- readRDS("../rObjects/dge.raw.all.samples.rds")
```

# Counts per million 
```{r cpm}
lcpm <- edgeR::cpm(dge$counts, log = TRUE)
cpm <- edgeR::cpm(dge$counts, log = FALSE)
write.table(cpm, "../results/all_samples/counts/cpm_raw.txt", quote = FALSE, sep = "\t")
write.table(lcpm, "../results/all_samples/counts/lcpm_raw.txt", quote = FALSE, sep = "\t")
```

# MDS
```{r MDS}
treat_colors <- c(treatment_colors)[dge$samples$Treatment]

par(bg = 'white')
plotMDS(
  lcpm,
  top = 100, 
  labels = dge$samples$Short.ID,
  cex = .8, 
  dim.plot = c(1,2), 
  plot = TRUE, 
  col = treat_colors,
  gene.selection = "common"
)
title(expression('Top 100 Genes (Log'[2]~'CPM)'))
saveToPDF("../results/all_samples/MDS/MDS_lcpm_100.pdf", width = 6, height = 5)


par(bg = 'white')
plotMDS(
  lcpm,
  top = 500, 
  labels = dge$samples$Short.ID,
  cex = .8, 
  dim.plot = c(1,2), 
  plot = TRUE, 
  col = treat_colors,
  gene.selection = "common"
)
title(expression('Top 500 Genes (Log'[2]~'CPM)'))
saveToPDF("../results/all_samples/MDS/MDS_lcpm_500.pdf", width = 6, height = 5)

par(bg = 'white')
plotMDS(
  cpm,
  top = 500, 
  labels = dge$samples$Short.ID,
  cex = .8, 
  dim.plot = c(1,2), 
  plot = TRUE, 
  col = treat_colors,
  gene.selection = "common"
)
title(expression('Top 500 Genes (Log'[2]~'CPM)'))
saveToPDF("../results/all_samples/MDS/MDS_cpm_500.pdf", width = 6, height = 5)
```

# JSD 
```{r JSD_heatmap}
# JDS
data <- JSD(t(edgeR::cpm(dge$counts)), est.prob = "empirical")
names <- dge$samples$Short.ID
colnames(data) <- names
rownames(data) <- names
round.data <- round(data, digits = 3) # round 3 decimal places

# set heatmap colors
colors <- c("blue","skyblue","white") # set heatmap color scale
colors <- colorRampPalette(colors)(1000) # make it a gradient

pheatmap::pheatmap(round.data,
                   treeheight_row = 0,
                   treeheight_col = 0,
                   display_numbers = TRUE,
                   fontsize_number = 9,
                   color = colors,
                   number_color = "black")

saveToPDF("../results/all_samples/MDS/JSD.pdf", width = 6, height = 4)

rm(data, names, colors, round.data)
```

# Sex check
```{r sex_check}
goi <- c("Uty","Ddx3y","Xist")

# create a list to store plots
plots_list <- list()

for (i in 1:length(goi)) {
  # get lcpm counts data 
  df <- lcpm[goi[i],]
  df <- reshape2::melt(df)
  df$sample <- rownames(df)
  
  p <- ggplot(data = df, mapping = aes(x = sample, y = value)) +
    geom_col() +
    theme_bw() +
    labs(title = goi[i], x = "sample", y = "lcpm")
  print(p)
  plots_list[[i]] <- p # add plots to list 
}
remove(df,p)

# arrange plots into a single facet-wrapped plot
facet_plot <- plot_grid(plotlist = plots_list, ncol = 1)
ggsave(filename = "../results/all_samples/goi/sex_check.pdf", plot = facet_plot)

rm(facet_plot, plots_list, goi, i)
```

# Contamination check
```{r contamination_check}
goi <- c("Ttr","Hbb-bs")

# create a list to store plots
plots_list <- list()

for (i in 1:length(goi)) {
  # get lcpm counts data 
  df <- lcpm[goi[i],]
  df <- reshape2::melt(df)
  df$sample <- rownames(df)
  
  p <- ggplot(data = df, mapping = aes(x = sample, y = value)) +
    geom_col() +
    theme_bw() +
    labs(title = goi[i], x = "sample", y = "lcpm")
  print(p)
  plots_list[[i]] <- p # add plots to list 
}
remove(df,p)

# arrange plots into a single facet-wrapped plot
facet_plot <- plot_grid(plotlist = plots_list, ncol = 1)
ggsave(filename = "../results/all_samples/goi/contamination_check.pdf", plot = facet_plot)

rm(facet_plot, plots_list, goi, i)
```

# Remove mitochondrial genes & keep only protein coding genes
```{r MT_genes}
dim(dge)
removeMT <- dge$genes$seqnames != "chrM"  # true when NOT MT
dge <- dge[removeMT,,keep.lib.sizes = FALSE]

keepProtein <- dge$genes$gene_type == "protein_coding"
dge <- dge[keepProtein,,keep.lib.sizes = FALSE]
dim(dge)

rm(keepProtein, removeMT)
```

# Library sizes
```{r library}
L <- mean(dge$samples$lib.size) * 1e-6
M <- median(dge$samples$lib.size) * 1e-6
c(L, M)
```

# Filter lowly expressed genes
The filterByExpr() function in the edgeR package determines which genes have a great enough count value to keep.  We will filter by group.  This means at least x samples (x is the smallest group sample size) must express a minimum count of 1 cpm. 
```{r filter}
keep.expr <- filterByExpr(dge, group = dge$samples$Treatment, min.count = M)
dim(dge)
dge.filtered <- dge[keep.expr, , keep.lib.sizes = FALSE]
dim(dge.filtered)
table(dge.filtered$genes$gene_type)

rm(L, M, keep.expr)
```

# TMM 
For estimating relative RNA production levels from RNA-seq data. 
The TMM method estimates scale factors between samples that can be incorporated into currently used statistical methods for DE analysis.
```{r TMM}
# Now, normalization by the method of trimmed mean of M-values (TMM)
dge.filtered.norm <- calcNormFactors(dge.filtered, method = "TMM")

# norm factor summary
summary(dge.filtered.norm$samples$norm.factors)
```

# Library density plot
Density plots of log-intensity distribution of each library can be superposed on a single graph for a better comparison between libraries and for identification of libraries with weird distribution. 
```{r library}
# set graphical parameter
par(mfrow = c(1,3))

# Normalize data for library size and expression intensity
log2cpm <- edgeR::cpm(dge, log = TRUE)
log2cpm.filtered <- edgeR::cpm(dge.filtered, log = TRUE)
log2cpm.norm <- edgeR::cpm(dge.filtered.norm, log = TRUE)

# set colors
colors <- treatment_colors[dge$samples$Treatment]
nsamples <- ncol(dge)

# First, plot the first column of the log2cpm density
plot(density(log2cpm[,1]), col = colors[1], lwd = 2, ylim = c(0,0.45), 
     las = 2, main = "A. Raw", xlab = expression('Log'[2]~CPM))

# For each sample plot the lcpm density
for (i in 2:nsamples){
  den <- density(log2cpm[,i]) #subset each column
  lines(den$x, den$y, col = colors[i], lwd = 2) 
}

# Second, plot log2cpm.filtered
plot(density(log2cpm.filtered[,1]), col = colors[1], lwd = 2, ylim = c(0,0.45), 
     las = 2, main = "B. Filtered", xlab = expression('Log'[2]~CPM))
abline(v = edgeR::cpm(3, log = TRUE), lty = 3)
for (i in 2:nsamples) {
  den <- density(log2cpm.filtered[,i])
  lines(den$x, den$y, col = colors[i], lwd = 2)
}

# Third, plot log2cpm.norm
plot(density(log2cpm.norm[,1]), col = colors[1], lwd = 2, ylim = c(0,0.45), 
     las = 2, main = "C. Normalized", xlab = expression('Log'[2]~CPM))
abline(v = edgeR::cpm(3, log = TRUE), lty = 3)
for (i in 2:nsamples) {
  den <- density(log2cpm.norm[,i])
  lines(den$x, den$y, col = colors[i], lwd = 2)
}
saveToPDF("../results/all_samples/library/gene_density.pdf", width = 6, height = 4)

rm(den, colors, i, nsamples, log2cpm, log2cpm.filtered, log2cpm.norm, dge.filtered, dge, cpm, lcpm)
```

# Save dge.filtered.norm R object
```{r}
saveRDS(dge.filtered.norm, file = "../rObjects/dge.filtered.norm.all.samples.rds")
```

# Add biomarker expression to dge metadata 
```{r biomarker_expression}
biomarkers <- c("Eno2", "Gfap", "Olig2", "Cd34", "P2ry12")
lcpm <- edgeR::cpm(dge.filtered.norm$counts, log = TRUE)

  for (i in biomarkers) {
  biomarker <- subset(genes.gtf, gene_name == i) # gene of interest 
  biomarker_counts <- subset(lcpm, rownames(lcpm) %in% biomarker)
  biomarker_melt <- reshape2::melt(biomarker_counts) # reshape data 
  # rename columns to merge with metadata 
  names(biomarker_melt)[names(biomarker_melt) == "value"] <- i 
  # rename columns to merge with metadata 
  names(biomarker_melt)[names(biomarker_melt) == "Var2"] <- "Short.ID" 
  names(biomarker_melt)[names(biomarker_melt) == "Var1"] <- "gene_id"
  biomarker_melt$gene_id <- NULL
  assign(paste0(i),biomarker_melt)
}

# add gene expression values into one dataframe 
# put all data frames into list
df_list <- list(Eno2, Gfap, Olig2, Cd34, P2ry12)

# merge all data frames in list
cell_biomarker_lcpm <- df_list %>% reduce(full_join, by='Short.ID')
write.table(cell_biomarker_lcpm, "../results/all_samples/counts/cell_biomarker_lcpm.txt", quote = FALSE, sep = "\t")

rm(biomarker, biomarker_counts, biomarker_melt, Cd34, Eno2, Gfap, Olig2, P2ry12, i, biomarkers, df_list)
```

# Scale data
rescaling a predictor in a regression has absolutely no effect on the magnitude of the relation being studiedâ€”the slope itself will not change its steepness, nor will the p-values or variance explained be changed. Rescaling is merely a means of communicating the nature of the regression line in different, hopefully more intuitive language.
```{r scale}
df <- merge(cell_biomarker_lcpm, metadata, by = "Short.ID")
scaled.info <-
  df[c("Age.at.harvest", "Eno2", "Gfap", "Olig2", "Cd34", "P2ry12")] %>% scale()
scaled.info.df <- as.data.frame(scaled.info)
# remove columns with unscaled data 
df <- (subset(df, select = -c(Age.at.harvest, Eno2, Gfap, Olig2, Cd34, P2ry12)))
# Add scaled information to the metadata called "info"
info_with_scale <- cbind(df, scaled.info.df)

rm(df, scaled.info.df, scaled.info, cell_biomarker_lcpm)
```


```{r add_scale_data_to_dge}
# Find the matching indices
match_indices <- match(dge.filtered.norm$samples$Short.ID, info_with_scale$Short.ID)
# Reorder info_with_scale using the matching indices
info_with_scale_reordered <- info_with_scale[match_indices, ]
all.equal(dge.filtered.norm$samples$Short.ID, as.character(info_with_scale_reordered$Short.ID))
# replace sample information with the updated info that includes biomakrer expression
dge.filtered.norm$samples <- info_with_scale_reordered
rownames(info_with_scale_reordered) <- info_with_scale_reordered$Short.ID
info <- info_with_scale_reordered
rm(info_with_scale, info_with_scale_reordered)
```

# Voom transform counts 
```{r voom, warning=FALSE}
counts = DGEList(counts = dge.filtered.norm$counts)
formula <- (~ 0 + Treatment)

# compute observation weights
vobj <- voomWithDreamWeights(counts, formula, metadata, plot = TRUE)

path <- paste0("../results/all_samples/voom/raw.voom")
saveToPDF(paste0(path, ".pdf"), width = 6, height = 4)

# voom counts
voomCounts <- vobj$E

rm(formula)
```

# CCA 
```{r CCA}
form <- ~ Treatment + Age.at.harvest + Eno2 + Gfap + Olig2 + Cd34 + Flowcell + Lane
# Compute Canonical Correlation Analysis (CCA) between all pairs of variables
# returns absolute correlation value
C = canCorPairs(form, info)
# Plot correlation matrix
plotCorrMatrix(C)

saveToPDF("../results/all_samples/variance/CCA.pdf", width = 8, height = 8)

rm(C, form)
```

# Fit variance 
variancePartition quantifies and interprets multiple sources of biological and technical variation in gene expression experiments. The package a linear mixed model to quantify variation in gene expression attributable to individual, tissue, time point, or technical variables.
```{r varpart}
form_varPart <- ~ (1|Treatment) + Age.at.harvest + Eno2 + Gfap + Cd34 + Olig2
varPart <- fitExtractVarPartModel(voomCounts, form_varPart, info)
vp <- sortCols(varPart)
```

```{r save_varpart}
saveRDS(vp, paste0("../rObjects/all_samples/variancePartition.rds"))
```

# Plot variance
```{r}
plotVarPart(vp)
saveToPDF("../results/all_samples/variance/variance_violins.pdf", width = 8.5, height = 5)

plotPercentBars(vp[1:10,])
saveToPDF("../results/all_samples/variance/variance_genes_top10_median.pdf", width = 7, height = 5)

write.table(vp, "../results/all_samples/variance/variance_explained.tsv", sep = "\t", quote = FALSE)

rm(vp)
```

# Top variance within Treatment
```{r top_var_treat}
# get gene with the highest variation across Treatment
# create data.frame with expression of gene i
i <- which.max(varPart$Treatment)
GE <- data.frame(
  Expression = voomCounts[i, ],
  Treatment = info$Treatment
)

# plot expression stratified by Treatment
label <- paste("Treatment:", format(varPart$Treatment[i] * 100,
  digits = 3
), "%")
main <- rownames(voomCounts)[i]
plotStratify(Expression ~ Treatment, GE,
  colorBy = NULL,
  text = label, main = main
)

saveToPDF("../results/all_samples/variance/variance_top_gene_treatment.pdf", width = 6, height = 4)
```

# Design matrix
```{r design}
design <-
  model.matrix(~ 0 + Treatment,
    dge.filtered.norm$samples
  )

colnames(design) <- c("Adu", "Fc", "IgG")
```

# Voom
When the library sizes are quite variable between samples, then the voom approach is theoretically more powerful than limma-trend. 
The voom method estimates the mean-variance relationship of the log-counts.
Generates a precision weight for each observation and enters these into the limma empirical Bayes analysis pipeline.
```{r voom_BIC}
form <- (~ 0 + Treatment)
#counts = DGEList(counts = dge.filtered.norm$counts)

voom_cov <-
  variancePartition::voomWithDreamWeights(
    counts = dge.filtered.norm$counts,
    formula = form,
    data = dge.filtered.norm$samples,
    BPPARAM = BiocParallel::SnowParam(cores),
    plot = TRUE
  )
voomCounts <- voom_cov$E
```

# Contrast plot
```{r contrasts}
# fits linear model for each gene given a series of arrays
fit <- lmFit(voom_cov, design)
coef.fit <- fit$coefficients

contrasts <- makeContrasts(
  Adu_vs_Fc = Adu - Fc,
  Adu_vs_IgG = Adu - IgG,
  Fc_vs_IgG = Fc - IgG,
  levels = colnames(design))
head(contrasts)

# save contrast names
allComparisons <- colnames(contrasts)
allComparisons # check

# run contrast analysis
vfit <- contrasts.fit(fit, contrasts = contrasts)

# Compute differential expression based on the empirical Bayes moderation of the
# standard errors towards a common value.
# The logCPM values can then be used in any standard limma pipeline, using the trend=TRUE
# argument when running eBayes or treat. For example:
veBayesFit <- eBayes(vfit, trend = TRUE, robust=TRUE)
plotSA(veBayesFit, main = "Final Model: Mean-variance Trend")

#saveToPDF(paste0(path, ".pdf"), width = 9, height = 5)
```

# DEGs summary
```{r DGE_summary}
pval <- 0.1
lfc.cutoff <- 0.0

sumTable <- 
  summary(decideTests(
    veBayesFit,  # object
    adjust.method = "BH", # by default the method = "separate"
    p.value = pval,
    lfc = lfc.cutoff  # numeric, minimum absolute log2-fold change required
  ))

print(paste0(" q < ", pval,
             " & absolute log2-fold change > ", lfc.cutoff))
sumTable
write.table(sumTable, "../results/all_samples/DEGs/DEGs.q0.1.summary.txt", quote = FALSE, sep = "\t")
```
# Add gene information to DEGs
reformat genes table to only include relevant information
```{r gene_relevant}
genes_relevant <- dplyr::select(genes.gtf, 1:4,10,12,13)
```

# Save objects
```{r save_voom}
saveRDS(veBayesFit, file = "../rObjects/all_samples/veBayesFit.rds")
saveRDS(voomCounts, file = "../rObjects/all_samples/voomCountsMatrix.rds")
```

# Output DEG tables
```{r DGE_output}
coef <- 1

for (i in allComparisons) {
  vTopTableAll <- topTable(
    veBayesFit, 
    coef = coef,  
    n = Inf, 
    p.value = 1,
    lfc = 0, 
    sort.by = "P", 
    genelist = genes_relevant, 
    confint = TRUE # column of confidence interval 
    )
    saveRDS(vTopTableAll, file = 
            paste0("../rObjects/all_samples/gene_tables/",
                   i,"_gene_table.rds"))
  path <- paste0("../results/all_samples/DEGs/", 
  i, "_gene_DEGs_q1.00.txt", sep = "") 
  write.table(
    vTopTableAll,
    path,
    sep = "\t",
    row.names = FALSE,
    quote = FALSE
  )
  # increment 
  coef <- coef + 1
}
remove(coef)
```

# PCA
Principal component analysis, or PCA, is a dimensionality reduction method that is often used to reduce the dimensionality of large data sets, by transforming a large set of variables into a smaller one that still contains most of the information in the large set.
```{r PCA}
# Setting the N of genes to use
ntop = length(dge.filtered.norm$genes$gene_id)
# Sorting by the coefficient of variance
means <- rowMeans(voomCounts)
Pvars <- rowVars(voomCounts, useNames = TRUE)
cv2 <- Pvars / means ^ 2
select <-
  order(cv2, decreasing = TRUE)[seq_len(min(ntop, length(cv2)))]
head(select)

highly_variable_exp <- ((voomCounts)[select,])
dim(highly_variable_exp)
# Running PCA
pca_exp <- prcomp(t(highly_variable_exp), scale = F, center = T)
# scale a logical value indicating whether the variables should be scaled
# to have unit variance before the analysis takes place.
# a logical value indicating whether the variables should be shifted to be zero centered.
head(pca_exp$x)[, 1:3]
summary(pca_exp)
# Dataframe with the first 9 PCs
dim1_10 <- data.frame(pca_exp$x[, 1:9])
# Adding metadata
dim1_10$Short.ID <- rownames(dim1_10)
pcaWithMetadata <- merge(dim1_10, info, by = "Short.ID", all = TRUE)
pcaWithMetadata$group <- pcaWithMetadata$Treatment

# Plotting
ggplot(data = pcaWithMetadata, aes(x = PC1, y = PC2, shape = group, color = group)) +
  geom_point(size = 2.5) +
  theme_bw()
```

# Info with PCA output 
```{r}
write.table(pcaWithMetadata, 
            paste0("../rObjects/all_samples/metadata.PCA.txt"), 
            quote = FALSE, sep = "\t")

```

# CCA PC1-10 & variables in model 
```{r CCA_PCA}
form_PCA <- ~ Treatment + 
  Age.at.harvest + 
  Eno2 +
  Gfap + 
  Cd34 + 
  Olig2 +
  PC1 +
  PC2 +
  PC3 +
  PC4 +
  PC5 +
  PC6 +
  PC7 +
  PC8 +
  PC9 

C = canCorPairs(form_PCA, pcaWithMetadata)
# Plot correlation matrix
cor.mtest <- function(C, ...) {
    C <- as.matrix(C)
    n <- ncol(C)
    p.mat<- matrix(NA, n, n)
    diag(p.mat) <- 0
    for (i in 1:(n - 1)) {
        for (j in (i + 1):n) {
            tmp <- cor.test(C[, i], C[, j], ...)
            p.mat[i, j] <- p.mat[j, i] <- tmp$p.value
        }
    }
  colnames(p.mat) <- rownames(p.mat) <- colnames(C)
  p.mat
}
# matrix of the p-value of the correlation
p.mat <- cor.mtest(C)
col <- colorRampPalette(c("#4477AA", "#77AADD", "#FFFFFF", "#EE9988", "#BB4444"))
  corrplot(C, method="color", col=col(200),  
         type="upper", order="hclust", 
         addCoef.col = "black", # Add coefficient of correlation
         tl.col="black", tl.srt=45, #Text label color and rotation
         diag=FALSE, col.lim = c(0, 1)
         )
saveToPDF("../results/all_samples/variance/corr_PCA.pdf", width = 10, height = 10)
```

# Volcano plots
```{r volcano}
for (i in allComparisons) {
  group1_vs_group2 <-
    read.delim(
      paste0(
        "../results/all_samples/DEGs/",i,
        "_gene_DEGs_q1.00.txt"
      )
    )
  color_values <- vector()
  max <- nrow(group1_vs_group2)
  for (row in 1:max) {
    if (group1_vs_group2$adj.P.Val[row] < 1) {
      if (group1_vs_group2$logFC [row] > 2) {
        color_values <- c(color_values, 1)
      }
      else if (group1_vs_group2$logFC[row] < -2) {
        color_values <- c(color_values, 2)
      }
      else {
        color_values <- c(color_values, 3)
      }
    }
    else{
      color_values <- c(color_values, 3)
    }
  }
  group1_vs_group2$color_adjpval_0.05 <- factor(color_values)
  data <- group1_vs_group2
  # plot only if there are DEGs with p_val_adj < 0.05
  num <- subset(data, (adj.P.Val < 1 & logFC < -2)  | (adj.P.Val < 1 & logFC > 2 ))
  num <- nrow(num)
  if (num != 0) {
    up <- data[data$color_adjpval_0.05 == 1,]
    up10 <- up[1:10,]
    upFold <- subset(up, logFC > 1)
    upFold <- upFold[!(upFold$gene_name %in% up10$gene_name),]
    down <- data[data$color_adjpval_0.05 == 2,]
    down10 <- down[1:10,]
    downFold <- subset(down, logFC < -1)
    downFold <- downFold[!(downFold$gene_name %in% down10$gene_name),]
    if (!1 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("blue", "gray")
    } else if (!2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("red", "gray")
    } else if (!1 %in% unique(data$color_adjpval_0.05) &&
               !2 %in% unique(data$color_adjpval_0.05)) {
      my_colors <- c("gray")
    } else {
      my_colors <- c("red", "blue", "gray")
    }
    hadjpval <- (-log10(max(data$P.Value[data$adj.P.Val < 0.05],
                            na.rm = TRUE)))
    negFC <- c(-2)
    posFC <- c(2) 
    i <- gsub("_vs_", " vs ", i)
    p <-
      ggplot(data = data,
             aes(
               x = logFC,
               y = -log10(P.Value),
               color = color_adjpval_0.05
             )) +
      geom_point(alpha = 0.8, size = 1) +
      theme_bw() +
      theme(legend.position = "none") +
      scale_color_manual(values = my_colors) +
      labs(
        title = "",
        x = expression(log[2](FC)),
        y = expression(-log[10] ~ "(" ~ italic("p") ~ "-value)")
      )  +
      geom_hline(yintercept = hadjpval,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = negFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      geom_vline(xintercept = posFC,
                 #  horizontal line
                 colour = "#000000",
                 linetype = "dashed") +
      ggtitle(paste0(i)) +
            geom_text_repel(
        data = up10,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "maroon",
        fontface = "italic",
        size = 3,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      ) +
      geom_text_repel(
        data = upFold,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "maroon",
        fontface = "italic",
        size = 3,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      ) +
      geom_text_repel(
        data = down10,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "navyblue",
        fontface = "italic",
        size = 3,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      ) +
      geom_text_repel(
        data = downFold,
        aes(
          x = logFC,
          y = -log10(P.Value),
          label = gene_name
        ),
        color = "navyblue",
        fontface = "italic",
        size = 3,
        max.overlaps = getOption("ggrepel.max.overlaps", default = 10)
      )
    p
    i <- gsub(" vs ", "_vs_", i)
    # save
    path <-
      paste0(
        "../results/all_samples/volcano/",
        i,
        "_gene_volcano_log2fc_2"
      )
    pdf(paste0(path, ".pdf"),
        height = 5,
        width = 5)
    print(p)
    dev.off()
    
    print(paste("i =", i))
  }
} 
```

```{r cleanup}
# clean up
remove(up, up10, upFold, group1_vs_group2, downFold, down10, data, p)
```

# Make excel table
```{r excel}
# first read in the files
for (i in allComparisons) {
  filepath <- paste0("../results/all_samples/DEGs/", i, "_gene_DEGs_q1.00.txt")
  assign(paste0(i),
         read.delim(
           filepath,
           header = TRUE,
           sep = "\t",
           stringsAsFactors = FALSE
         ))
}
allComparisons
list_of_datasets <- list(
"Adu_vs_Fc" = Adu_vs_Fc,
"Adu_vs_IgG" = Adu_vs_IgG,     
"Fc_vs_IgG" = Fc_vs_IgG)
write.xlsx(list_of_datasets, file = paste0("../results/all_samples/DEGs/DEGs.FDRq1.00.xlsx"))
```

```{r}
sessionInfo()
```